src/plugins/generic/sqlitehistory/sqlitehistory.cpp

```diff
@@ -0,0 +1,506 @@
+	query.bindValue(":key", "sqlitehistory");
+	query.exec();
+
+	if(query.first()) {
+		int version = query.value(0).toInt();
+		if(version == currentVersion()) {
+			qDebug() << "No migration needed";
+			return;
+		} else if(version > currentVersion()) {
+			qDebug() << "Current version of sqlitehistory plugin is" << currentVersion();
+			qDebug() << "But qutim.sqlite version = " << version;
+			qFatal("qutIM sqlite database is older than plugin. Cannot proceed. Please upgrade qutIM to last version");
+		} else if(version < currentVersion()) {
+			qDebug() << "Database older than plugin, executing migrations";
+			for(int i = version + 1; i < currentVersion(); ++i)
+				makeMigration(i);
+		}
+	} else {
+		qDebug() << "First run, inserting current version";
+		QSqlQuery query;
+
+		query.prepare("INSERT INTO qutim_history_version (key, value) "
+					  "VALUES (:key, :value)");
+		query.bindValue(":key", "sqlitehistory");
+		query.bindValue(":value", currentVersion());
+		query.exec();
+
+		qDebug() << "No migration needed";
+	}
+}
+
+void SqliteWorker::exec()
+{
+	forever {
+		m_queueLock.lock();
+		if(m_queue.isEmpty()) {
+			m_runningLock.lock();
+			m_isRunning = false;
+			m_runningLock.unlock();
+
+			m_queueLock.unlock();
+			break;
+		}
+
+		auto f = m_queue.dequeue();
+		m_queueLock.unlock();
+
+		f();
+	}
+}
+
+void SqliteWorker::runJob(std::function<void ()> job)
+{
+	m_queueLock.lock();
+	m_queue.enqueue(std::move(job));
+	m_queueLock.unlock();
+
+	m_runningLock.lock();
+	if(!m_isRunning) {
+		m_isRunning = true;
+		m_runningLock.unlock();
+		exec();
+	}
+}
+
+/**
+ * Escapes string to use in LIKE query. You need to add ESCAPE '@' to end of your query
+ */
+QString SqliteWorker::escapeSqliteLike(const QString &str)
+{
+	QString escaped = str;
+	escaped.replace('@', "@@");
+	escaped.replace('_', "@_");
+	escaped.replace('%', "@%");
+
+	return escaped;
+}
+
+void SqliteWorker::process()
+{
+	m_db = QSqlDatabase::addDatabase("QSQLITE");
+	QDir history_dir = SystemInfo::getDir(SystemInfo::HistoryDir);
+	QString pathToHistory = history_dir.absolutePath() + QDir::separator() + "qutim.sqlite";
+	qDebug() << "Trying to open database" << pathToHistory;
+	m_db.setDatabaseName(pathToHistory);
+	bool openSuccess = m_db.open();
+
+	if(!openSuccess) {
+		qDebug() << m_db.lastError();
+		qFatal("Cannot open sqlite database");
+	} else
+		qDebug() << "Database opened!";
+
+	prepareDb();
+
+	m_runningLock.lock();
+	m_isRunning = true;
+	m_runningLock.unlock();
+	exec();
+}
+
+void SqliteWorker::makeMigration(int version)
+{
+	qDebug() << "Executing migration to version" << version;
+	m_db.transaction();
+	QSqlQuery query;
+
+	switch(version) {
+	case 0:
+		query.prepare("ALTER TABLE qutim_history ADD COLUMN `sendername` TEXT DEFAULT NULL");
+		break;
+	default:
+		qFatal("Unhandled migration to version %i! Seems like a bug", version);
+		break;
+	}
+
+	query.exec();
+
+	m_db.commit();
```

Comment:
Shouldn't we do commit after checking if query failed or not? Also, I believe, we should do qutim_history_version update inside transaction as well and do rollback if either of them failed.